janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -732,6 +757,7 @@ public JanusGraphVertexProperty addProperty(VertexProperty.Cardinality cardinali
         for (IndexLockTuple lockTuple : uniqueIndexTuples) uniqueLock = new CombinerLock(uniqueLock,getLock(lockTuple),times);
         uniqueLock.lock(LOCK_TIMEOUT);
         try {
+            checkPropertyConstraintForVertexOrCreateConstraint(vertex.vertexLabel(), key);
```

Comment:
I think that this shouldn't be here as this part checks for uniqueness constraints and even uses a lock for that. So this should either be before the code responsible for the unique constraints or after that.